File: temp.java
#include <bits/stdc++.h>
using namespace std;

void unionFind(int node, unordered_map<int, int> &par)
{
    if (par[node] == node)
        return;
    unionFind(par[node], par);
    par[node] = par[par[node]];
}

void updateGroup(int u, int v, unordered_map<int, int> &par, unordered_map<int, int> &group, int count[2])
{
    unionFind(v, par);
    unionFind(u, par);

    if (par[u] == par[v])
        return;

    int oldOdd = 0, oldEven = 0, newOdd = 0, newEven = 0;

    if (group[par[u]] != 1 && group[par[u]] % 2)
        oldOdd++;

    if (group[par[u]] != 1 && group[par[u]] % 2 == 0)
        oldEven++;

    if (group[par[v]] != 1 && group[par[v]] % 2)
        oldOdd++;

    if (group[par[v]] != 1 && group[par[v]] % 2 == 0)
        oldEven++;

    group[par[v]] = group[par[v]] + group[par[u]];

    par[u] = par[v];
    unionFind(v, par);
    unionFind(u, par);

    if (par[u] != par[v] && group[par[u]] != 1 && group[par[u]] % 2)
        newOdd++;

    if (par[u] != par[v] && group[par[u]] != 1 && group[par[u]] % 2 == 0)
        newEven++;

    if (group[par[v]] != 1 && group[par[v]] % 2)
        newOdd++;

    if (group[par[v]] != 1 && group[par[v]] % 2 == 0)
        newEven++;

    count[0] = count[0] - oldOdd + newOdd;
    count[1] = count[1] - oldEven + newEven;

}
int32_t main()
{
    int t, i, j, n, m, itemp;
    cin >> t;
    for (auto tc = 1; tc <= t; tc++)
    {
        cin >> n;
        vector<int> v(n);
        unordered_map<int, vector<int>> al;
        unordered_map<int, int> par, group;
        for (int i = 1; i <= n; i++)
        {
            par[i] = i;
            group[i] = 1;
        }

        int count[2] = {0};

        for (int inp = 1; inp < n; inp++)
        {
            int u, v;
            cin >> u >> v;

            al[u].push_back(v);
            al[v].push_back(u);

            updateGroup(u, v, par, group, count);

            cout << abs(count[1] - count[0]) << ' ';
        }
        cout << endl;
    }
    return 0;
}

File: temp.cpp
#include <bits/stdc++.h>
using namespace std;

bool is_adjacent(char A, char B)
{
    return (abs(A - B) == 1); // 'A' 'B' and 'B' 'A' etc
}


void move_disks(int n, char source, char destination)
{
    if (n == 0 || source == destination)
    {
        return;
    }
    if (is_adjacent(source, destination))
    {
        char auxiliary;
        if ((source == 'A' && destination == 'B') || (source == 'B' && destination == 'A'))
        {
            auxiliary = 'C';
        }
        else
        {
            auxiliary = 'A';
        }
        move_disks(n - 1, source, auxiliary);
        cout << "Move disk " << n << " from " << source << " to " << destination << endl;
        move_disks(n - 1, auxiliary, destination);
    }
    else
    {
        move_disks(n - 1, source, destination);
        cout << "Move disk " << n << " from " << source << " to B" << endl;
        move_disks(n - 1, destination, source);
        cout << "Move disk " << n << " from B to " << destination << endl;
        move_disks(n - 1, source, destination);
    }
}
int main()
{
    int n;
    cin >> n;
    move_disks(n, 'A', 'B');
    return 0;
}

/*
Move disk 1 from A to B
Move disk 1 from B to C
Move disk 2 from A to B
Move disk 1 from C to B
Move disk 1 from B to A
Move disk 2 from B to C
Move disk 1 from A to B
Move disk 1 from B to C
Move disk 3 from A to B
Move disk 1 from C to B
Move disk 1 from B to A
Move disk 2 from C to B
Move disk 1 from A to B

Move disk 1 from B to A
Move disk 2 from B to C
Move disk 1 from A to B
Move disk 1 from B to C
Move disk 3 from B to A
Move disk 1 from C to B
Move disk 1 from B to A
Move disk 2 from C to B
Move disk 1 from A to B
Move disk 1 from B to C
Move disk 2 from B to A
Move disk 1 from C to B
Move disk 1 from B to A
*/